### ISSUE TITLE
KVM vmrun failed with invalid argument

### ISSUE THREAD

**Issue by mincheolsung** [AMDESE/AMDSEV#131]
Hi,

I tried to suspend and resume a SEV-ES VM running with KVM.
The processes are:
1. The QEMU monitor stops the VM with "stop" command
2. The QEMU monitor calls SEND_START, SEND_UPDATE_VMSA, SEND_FINISH
3. The QEMU monitor calls RECEIVE_START (with new ASID and SEV handle), RECEIVE_UPDATE_VMSA, RECEIVE_FINISH
4. The KVM switches VM's vmcb's vmsa_pa with the new address of the new VMSA imported by RECEIVE_UPDATE_VMSA command.
5. The QEMU monitor resumes the VM with "cont" command

However, the VM fails to resume with "kvm run failed Invalid argument". This meant VMRUN failure.
I know the entire memory is encrypted with the old ASID but I guess the VM will crash even after a successful VMRUN.
I actually tried this without SEND_* and RECEIVE_* cmmands and just initialized a new VMSA with initial states. So the vCPU went to the initial state which was the real mode.

Does anyone have any idea why the VMRUN failed?

Another question is wheter I should set the C-bit when replacing VMCB's vmsa_pa with the new address.
I am wondering since I set the C-bit when calling SEND_UPDATE_VMSA and RECEIVE_UPDATE_VMSA.

**Comment by tlendacky** [comment]
If the CPU can't correctly decrypt the memory, you'll likely take a fault, which will try to follow the IDT, which it also won't be able to decrypt properly, etc...  so you're likely triple faulting. What's the value of the VMCB exit_int_info field for the vCPU (echo 1 > /sys/module/kvm_amd/parameters/dump_invalid_vmcb) on failure?

**Comment by mincheolsung** [comment]
I got exit_code: 7f, info1: 0, info2: 0.

**Comment by tlendacky** [comment]
The dump is located in the dmesg output. The exit_int_info field is at offset 0x88 in the VMCB if you prefer to just print that.

**Comment by mincheolsung** [comment]
@tlendacky It was 0x80000b08.

**Comment by tlendacky** [comment]
> It was 0x80000b08.

That is indicative of a triple fault. The x08 in the LSB indicates that the vCPU was trying to handle a double fault when the exit occurred. Since it was a SHUTDOWN exit code, that typically means the double fault couldn't be handled and the processor triple faulted.

Exceptions aren't interrupts, so the IF flag won't prevent this. You basically have nothing that the VMRUN can execute. Until you get the memory into the new key, your VMRUN is likely to always "fail". Note, that you didn't get a VMEXIT_INVALID exit code, which means that your VMSA appears to be valid, it just can't fetch proper instructions or handle exceptions from bad instructions.

**Comment by mincheolsung** [comment]
@tlendacky Thanks for the explanation!
I have the following question. Can I make VMRUN to execute unencrypted code by setting the nRIP?

**Comment by RARelph** [comment]
No. An SEV guest will implicitly set the C-bit for all instruction fetches and MMU table walks.

**Comment by mincheolsung** [comment]
What if copy and reencrypt code with SEND_UPDATE_DATA?

**Comment by RARelph** [comment]
That's required. You need to move the guest data using SEND_UPDATE_DATA and RECEIVE_UPDATE_DATA.

**Comment by mincheolsung** [comment]
My final question is about nRIP.
I just want to make sure I can let VMRUN execute code pointed by nRIP.

**Comment by tlendacky** [comment]
No, nRIP cannot be used to do that, it defeats the purpose of SEV-ES / SEV-SNP. Register state cannot be changed by the hypervisor.

**Comment by tlendacky** [comment]
No!  What prevents the hypervisor from setting nRIP after that?  nRIP is in the VMCB, not the VMSA.

**Comment by mincheolsung** [comment]
Thanks. I will close this. If you don't mind, can I ask just a few more questions by email?
I don't want to disclose them.

**Comment by mincheolsung** [comment]
I am wondering how the guest executes reencrypted code after VMRUN.
My thought is:
1. Suspend the guest just before executing code at gpa "a".
2. Hypervisor re-encrypts the code and remaps NPT to map the gpa "a" to the re-encrypted code.
3. Then the guest executes the re-encrypted code at the gpa "a".

Are these feasible?

**Comment by RARelph** [comment]
Every guest instruction fetch and guest page table walk will use the hardware's decryption path using that specific guest's memory encryption key. This is unavoidable. The only source of properly encrypted data for a guest is a) the LAUNCH_UPDATE_DATA or RECEIVE_UPDATE_DATA commands; or b) the guest itself by writing the data with the C-bit set. That's it. We've been over this on your previous ticket.

What are you trying to do?

**Comment by mincheolsung** [comment]
I forgot to say I created a guest page table on a decrypted memory (where C-bit is not set).
I just want to switch the encryption key and execute a simple code (even just few instructions).

**Comment by tlendacky** [comment]
If you mean a guest page table (as in the CR3 register of the guest), then it has to be encrypted. As @RARelph said, every guest page table walk will use the hardware's decryption path. So creating a guest page table in unencrypted memory won't work. Now if you're referring to the nested page table that is maintained by the hypervisor, that is a different story.

**Comment by mincheolsung** [comment]
Ok, then I can also reencrypt the guest page table with SEND_* and RECEIVE_* commands.
What about other parts?

> I am wondering how the guest executes reencrypted code after VMRUN. My thought is:
> 
> 1. Suspend the guest just before executing code at gpa "a".
> 2. Hypervisor re-encrypts the code and remaps NPT to map the gpa "a" to the re-encrypted code.
> 3. Then the guest executes the re-encrypted code at the gpa "a".
> 
> Are these feasible?

**Comment by tlendacky** [comment]
Yes, think about how live migration would be performed and that is basically what you will need to do. Any pages that are encrypted will need to be go through the SEND_* and RECEIVE_* flow. Pages that are not encrypted would need to be copied as is. Then you need to map those new migrated pages in a new nested page table. You will need everything that may be accessed during a VMRUN in the correct state (IDT, GDT, etc.), otherwise you risk a SHUTDOWN exit.
